include/lin/substitutions.hpp

```diff
@@ -1,11 +1,273 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
 /** @file lin/substitutions.hpp
  *  @author Kyle Krol
- *  Collection of header files composing the substitutions module. */
+ */
+
+/** @defgroup SUBSTITUTIONS Substitutions
+ *
+ *  @brief Supports solving linear system in conjunction with matrix
+ *         factorizations.
+ */
 
 #ifndef LIN_SUBSTITUTIONS_HPP_
 #define LIN_SUBSTITUTIONS_HPP_
 
-#include "substitutions/backward_substitution.hpp"
-#include "substitutions/forward_substitution.hpp"
+#include "core.hpp"
+#include "references.hpp"
+
+#include <type_traits>
+
+namespace lin {
+namespace internal {
+
+/** @internal
+ *
+ *  @brief Tests whether or not the backward substitution algorithm can be
+ *         applied to a set of types.
+ *
+ *  @tparam C Upper triangular operator type.
+ *  @tparam D Unknown matrix or vector type.
+ *  @tparam E Know matrix or vector type.
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E>
+struct can_backward_sub : conjunction<
+    can_multiply<C, D>, have_same_elements<C, D, E>, have_same_dimensions<D, E>,
+    is_square<C>
+  > {};
+
+/** @internal
+ *
+ *  @brief Tests whether or not the forward substitution algorithm can be
+ *         applied to a set of types.
+ *
+ *  @tparam C Lower triangular operator type.
+ *  @tparam D Unknown matrix or vector type.
+ *  @tparam E Know matrix or vector type.
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E>
+struct can_forward_sub : can_backward_sub<C, D, E> {};
+
+} // namespace internal
+
+/** @brief Solves a linear system using backward substitution.
+ *
+ *  @tparam C
+ *  @tparam D
+ *  @tparam E
+ *
+ *  @param U Upper triangular operator.
+ *  @param X Unknown matrix or vector.
+ *  @param Y Know matrix or vector.
+ *
+ *  @return Zero on success and some other value otherwise.
+ *
+ *  Solves a linear system of the following form:
+ *
+ *  \f[
+ *    U X = Y
+ *  \f]
+ * 
+ *  where \f$U\f$ is an upper triangular operator, \f$X\f$ is an unknow matrix
+ *  or vector, and \f$Y\f$ is a known matrix or vector.
+ *
+ *  In order for the backward substitution to be valid, the matrix `U` must be
+ *  in upper triangular form. The function itself does not check for this and
+ *  simply assumes the elements below the main diagonal are all zero.
+ *
+ *  Lin assertion errors will be thrown if the dimensions of `X` are not the
+ *  same as the dimensions of `Y`.
+ *
+ *  @sa internal::can_backward_sub
+ *  @sa forward_sub
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E, typename =
+    std::enable_if_t<internal::can_backward_sub<C, D, E>::value>>
+constexpr size_t backward_sub(internal::Mapping<C> const &U, internal::Mapping<D> &X, internal::Mapping<E> const &Y) {
+  LIN_ASSERT(U.rows() == U.cols());
+  LIN_ASSERT(U.cols() == Y.rows());
```

Comment:
X.rows()?

Reply:
I mean yeah it could be that, but we also check that `X` and `Y` have the same dimensions so it works out. I had just copied this over from the old implementation lol